### PR TITLE
Add Auth Provider

### PR DESCRIPTION
**Pull request by lifekaizen**
**For discussion only - not ready to merge!**

Auth provider Interface and implementation with SubtleCrypto.

Addresses:
* #3 
* #7 
* #9

### CODE REVIEW

File: packages/collabswarm/src/auth-subtlecrypto.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+   * @remarks
+   * This is not Node’s Crypto API; that API is not expected to be as performant.
+   *
+   * @param _nonce_bits - 96 bits length is recommended in docs; though example uses only 12
+   */
+  _nonce_bits = 96; // TODO (e: Robert) or we can just force it, but then it's hidden if change algo
+
+  // Given encrypted changes and a private key,
+  // return a signature for use in a CRDTChangeBlock
+  // TODO (e:Robert) should we call it CRDTChangeBlock? esp since sig is optional
```
Comment (robotoer) [inline packages/collabswarm/src/auth-subtlecrypto.ts:18]:
```suggestion
```

File: packages/collabswarm/src/auth-subtlecrypto.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  /**
+   * Given encrypted data, key and a nonce,
+   * return the decrypted data or throw an error
+   *
+   * @remarks
+   * Expects that nonce has been separated out from data
+   *
+   * @param data - encrypted data as uint_8 array, not including nonce
+   * @param documentKey - symmetric key associated and stored with document
+   * TODO (e:Robert) - documentKey should become documentKeys since it will change?
```
Comment (robotoer) [inline packages/collabswarm/src/auth-subtlecrypto.ts:62]:
```suggestion
```

We can make a `tryDecrypt` that tries multiple keys

File: packages/collabswarm/src/collabswarm-document.ts [hunk 3]
```diff
@@ -35,6 +35,8 @@ export class CollabswarmDocument<DocType, ChangesType, ChangeFnType, MessageType
     private readonly _provider: CRDTProvider<DocType, ChangesType, ChangeFnType, MessageType>,
     private readonly _changesSerializer: ChangesSerializer<ChangesType>,
     private readonly _messageSerializer: MessageSerializer<MessageType>,
+    private _read_ACL: {},
+    private _write_ACL: {}
```
Comment (robotoer) [inline packages/collabswarm/src/collabswarm-document.ts:39]:
```suggestion
    private _writeACL: {[name: string]: CryptoKey}  // Or something like that.
```

The current typedef (`{}` as a type means specifically an empty dictionary only) prohibits adding any keys.

File: .vscode/settings.json [hunk 4]
```diff
@@ -1,4 +1,20 @@
...
+        "crdt",
+        "ipfs",
+        "libp",
+        "p",
+        "pss",
+        "tsc",
+        "websockets",
+        "wss"
+    ]
+>>>>>>> Stashed changes
```
Comment (robotoer) [inline .vscode/settings.json:19]: Looks like this merge conflict didn't get resolved properly

Comment (lifekaizen) [reply .vscode/settings.json]: Good catch!

**Comment by lifekaizen** [comment]
Mind if I turn on "format on save with prettier" so as to reduce the bot commits?

**Comment by robotoer** [comment]
Go for it!

File: packages/collabswarm-subtlecrypto/package.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
+{
```
Comment (robotoer) [inline packages/collabswarm-subtlecrypto/package.json:1]: @lifekaizen Is this package/module still used?

File: packages/collabswarm/src/auth-subtlecrypto.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+          name: this._encryptionAlgorithmName,
+          iv: iv_value,
+        };
+      default:
+        throw "Encryption is only supported with AesGcmParams currently"!;
+    }
+  }
+
+  // Given encrypted changes and a private key,
+  // return a signature for use in a CRDTChangeBlock
```
Comment (robotoer) [inline packages/collabswarm/src/auth-subtlecrypto.ts:74]: Nit: Update this to use a tsdoc comment

File: packages/collabswarm/src/auth-subtlecrypto.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    data: Uint8Array,
+    privateKey: CryptoKey
+  ): Promise<Uint8Array> {
+    return new Uint8Array(
+      await crypto.subtle.sign(this.signingAlgorithm, privateKey, data)
+    );
+  }
+
+  // Given a signature and data (from a CRDTChangeBlock),
+  // return a Promise that fulfills with true if the signature is valid, false otherwise
```
Comment (robotoer) [inline packages/collabswarm/src/auth-subtlecrypto.ts:85]: Nit: Update this to use a tsdoc comment

File: packages/collabswarm/src/utils-test.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
+// https://nodejs.org/api/webcrypto.html#webcrypto_exporting_and_importing_keys
```
Comment (robotoer) [inline packages/collabswarm/src/utils-test.js:1]: This file would be nice to have as typescript instead of javascript

Comment (robotoer) [reply packages/collabswarm/src/utils-test.js:1]: Is this still used?

File: packages/collabswarm-subtlecrypto/package.json (hunk 5 above)
Comment (lifekaizen) [reply packages/collabswarm-subtlecrypto/package.json:1]: Will remove!

File: packages/collabswarm/src/auth-subtlecrypto.ts (hunk 6 above)
Comment (lifekaizen) [reply packages/collabswarm/src/auth-subtlecrypto.ts:74]: Done

**Review by robotoer — APPROVED** [review]
LGTM!

🚂 🚋 🚋

**Pull request merged**